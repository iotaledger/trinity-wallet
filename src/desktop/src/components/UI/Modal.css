.backdrop {
    display: flex;
    align-items: center;
    justify-content: center;
    right: 0;
    top: 0;
    width: 100vw;
    height: 100vh;
    position: fixed;
    z-index: var(--z-index-modal);
    &:before {
        content: '';
        position: absolute;
        width: 100vw;
        height: 100vh;
        top: 0px;
        left: 0px;
        z-index: 1;
        background: var(--body-bg);
    }
    &.confirm {
        &:before {
            opacity: 0.95;
        }
        text-align: center;
        h1 {
            font-weight: 200;
            margin-bottom: 20px;
            max-width: none;
        }
        p {
            max-width: none;
        }
        :global {
            h1.error {
                color: var(--negative);
            }
            h1.success {
                color: var(--positive);
            }
        }
        strong {
            max-width: 40vw;
            word-wrap: break-word;
        }
        button {
            margin: 20px;
        }
    }
}

.wrapper {
    width: 80vw;
    max-width: 960px;
    position: relative;
    z-index: 2;
}
