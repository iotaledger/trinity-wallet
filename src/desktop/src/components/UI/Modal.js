import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import css from './Modal.css';

export default class Modal extends React.Component {
    static propTypes = {
        children: PropTypes.node,
        className: PropTypes.string,
        hideCloseButton: PropTypes.bool,
        isOpen: PropTypes.bool,
        onOpen: PropTypes.func,
        onClose: PropTypes.func,
        onStateChange: PropTypes.func,
    };

    state = {
        isOpen: this.props.isOpen,
    };

    componentWillReceiveProps(nextProps) {
        if (nextProps.isOpen !== this.state.isOpen) {
            this.toggle(nextProps.isOpen);
        }
    }

    // TODO: add an keyDown [esc] == close function here
    onOpen = () => {
        if (typeof this.props.onOpen === 'function') {
            return this.props.onOpen();
        }
    };

    onClose = () => {
        if (typeof this.props.onClose === 'function') {
            return this.props.onClose();
        }
    };

    onStateChange = newState => {
        if (typeof this.props.onStateChange === 'function') {
            return this.props.onStateChange(newState);
        }
    };

    toggle = isOpen => {
        const doneCallback = () => {
            this.onStateChange(this.state.isOpen);
            this.state.isOpen ? this.onOpen() : this.onClose();
        };
        if (isOpen === true || isOpen === false) {
            return this.setState(() => {
                return {
                    isOpen: isOpen,
                };
            }, doneCallback);
        }
        return this.setState(
            state => ({
                isOpen: !state.isOpen,
            }),
            doneCallback,
        );
    };

    close = () => {
        this.setState(
            () => ({
                isOpen: false,
            }),
            () => {
                this.onStateChange(false);
                this.onClose();
            },
        );
    };

    render() {
        const { isOpen } = this.state;
        const { className, hideCloseButton } = this.props;

        if (!isOpen) {
            return null;
        }

        return ReactDOM.createPortal(
            <div
                ref={node => {
                    this.backdropEl = node;
                }}
                className={classNames(css.backdrop, className)}
                onClick={e => e.target === this.backdropEl && this.close()}
            >
                <div className={css.wrapper}>
                    {!hideCloseButton && <button onClick={this.close}>X</button>}
                    <div className={css.content}>{this.props.children}</div>
                </div>
            </div>,
            document.getElementById('modal'),
        );
    }
}
