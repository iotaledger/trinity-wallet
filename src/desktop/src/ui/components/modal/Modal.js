import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import css from './modal.css';

/**
 * Modal window component
 */
export default class Modal extends React.Component {
    static propTypes = {
        /** Modal content */
        children: PropTypes.oneOfType([PropTypes.node, PropTypes.array]),
        /** Modal window type */
        variant: PropTypes.oneOf(['confirm']),
        /** Modal visibility state */
        isOpen: PropTypes.bool,
        /** Should the dialog be without a cancel option */
        isForced: PropTypes.bool,
        /** Modal inline style state */
        inline: PropTypes.bool,
        /** Modal visibility state */
        onClose: PropTypes.func.isRequired,
    };

    componentDidMount() {
        window.addEventListener('keydown', this.onKeyDown, false);
    }

    componentWillUnmount() {
        window.removeEventListener('keydown', this.onKeyDown, false);
    }

    onKeyDown = (e) => {
        if (e.key === 'Escape' && this.props.isOpen && !this.props.isForced) {
            this.props.onClose();
        }
    };

    render() {
        const { variant, isOpen, inline } = this.props;

        const content = (
            <div className={classNames(css.backdrop, css[variant], !isOpen ? css.hidden : null)}>
                <div className={css.wrapper}>
                    <div className={css.content}>{this.props.children}</div>
                </div>
            </div>
        );

        return inline ? content : ReactDOM.createPortal(content, document.getElementById('modal'));
    }
}
