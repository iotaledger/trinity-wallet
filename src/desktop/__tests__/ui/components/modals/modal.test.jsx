import React from 'react';
import { shallow } from 'enzyme';

import Modal from 'ui/components/modal/Modal';

const props = {
    children: 'Lorem ipsum',
    onClose: jest.fn(),
    inline: true,
    isOpen: false,
};

describe('Modal component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Modal {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Display content', () => {
        const wrapper = shallow(<Modal {...props} />);

        expect(wrapper.find('.content').text()).toEqual('Lorem ipsum');
    });

    test('Modal hidden', () => {
        const wrapper = shallow(<Modal {...props} />);

        expect(wrapper.hasClass('hidden')).toBeTruthy();
    });

    test('Modal visible', () => {
        const mockProps = Object.assign({}, props, { isOpen: true });
        const wrapper = shallow(<Modal {...mockProps} />);

        expect(wrapper.hasClass('hidden')).toBeFalsy();
    });
});
