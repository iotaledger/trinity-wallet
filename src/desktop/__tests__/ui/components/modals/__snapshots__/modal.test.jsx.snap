// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Render the component 1`] = `
<div
  className="backdrop undefined hidden"
>
  <div
    className="wrapper"
  >
    <div
      className="content"
    >
      Lorem ipsum
    </div>
  </div>
</div>
`;
